tests/test_router_direct_custom_quote_in_given_out.py

```diff
@@ -0,0 +1,87 @@
+import time
+import pytest
+
+import conftest
+from sqs_service import *
+from quote import *
+from quote_response import *
+from rand_util import *
+from e2e_math import *
+from decimal import *
+from constants import *
+from util import *
+from route import *
+
+# Arbitrary choice based on performance at the time of test writing
+EXPECTED_LATENCY_UPPER_BOUND_MS = 15000
+
+# Test suite for the /router/custom-direct-quote endpoint
+# Test runs tests for exact amount out quotes.
+class TestExactAmountOutDirectCustomQuote:
+    @pytest.mark.parametrize("pair", conftest.create_coins_from_pairs(conftest.create_no_dupl_token_pairs(conftest.choose_tokens_liq_range(num_tokens=10, min_liq=500_000, exponent_filter=USDC_PRECISION)), USDC_PRECISION, USDC_PRECISION + 3), ids=id_from_swap_pair)
+    def test_get_custom_direct_quote(self, environment_url, pair):
+        denom_in = pair['out_denom']
+        coin = Coin(pair['token_in']['denom'], pair['token_in']['amount_str'])
+        token_out  = pair['token_in']['amount_str'] + pair['token_in']['denom']
+
+        # Get the optimal quote for the given token pair
+        # Direct custom quote does not support multiple routes, so we request single/multi hop pool routes only
+        optimal_quote = ExactAmountOutQuote.run_quote_test(environment_url, token_out, denom_in, False, True, EXPECTED_LATENCY_UPPER_BOUND_MS)
+
+        pool_id = ','.join(map(str, optimal_quote.get_pool_ids()))
+        denoms_in = ','.join(map(str, optimal_quote.get_token_in_denoms()))
+
+        quote = self.run_quote_test(environment_url, token_out, denoms_in, pool_id, EXPECTED_LATENCY_UPPER_BOUND_MS)
```

Comment:
Great to see that these tests are dynamic.

However, could we add the hardcoded tests similar to "out given in" still?
https://vscode.dev/github/osmosis-labs/sqs/blob/SQS-303-e2e/tests/test_router_quote_out_given_in.py#L101-L146

Let's then wire **all** hardcoded test cases into our synthetic monitoring suite here 🙏 :
https://vscode.dev/github/osmosis-labs/sqs/blob/SQS-303-e2e/tests/test_synthetic_geo.py#L18

The synthetic monitoring suite runs every 20 minutes on schedule against every environment. As a result, we need more determinism via hardcoding so that those are not flaky.

Reply:
Synth tests from now on [includes](https://github.com/osmosis-labs/sqs/pull/423/files#diff-a08edff05b996a639012f732ff5db3c8f14bd6fc37c39c6303c531917211167dR70) test cases for given in/out quotes as well as for in/out custom direct quotes. :+1: But I am not sure I get idea of "all" right, if not, would creating a separate task would make more sense? Given that we quite improved tests state in overall with this PR by including all the rest tests to synh would be a bit out of the scope of original task and do not provide enough visibility.